quickjs.c

```diff
@@ -602,10 +616,22 @@ static int rebuild_ic(JSContext *ctx, JSInlineCache *ic);
 static int resize_ic_hash(JSContext *ctx, JSInlineCache *ic);
 static int free_ic(JSRuntime *rt, JSInlineCache *ic);
 static uint32_t add_ic_slot(JSContext *ctx, JSInlineCache *ic, JSAtom atom, JSObject *object,
-                            uint32_t prop_offset);
+                            uint32_t prop_offset, JSObject* prototype);
+
+static int ic_delete_shape_proto_watchpoints(JSRuntime *rt, JSShape *shape, JSAtom atom);
+static int ic_free_shape_proto_watchpoints(JSRuntime *rt, JSShape *shape);
 
-static int32_t get_ic_prop_offset(JSInlineCache *ic, uint32_t cache_offset,
-                                  JSShape *shape)
+static int ic_watchpoint_delete_handler(JSRuntime* rt, intptr_t ref, uint8_t offset, JSAtom atom, intptr_t target);
+static int ic_watchpoint_free_handler(JSRuntime* rt, intptr_t ref, uint8_t offset, JSAtom atom);
```

Comment:
```suggestion
static void ic_watchpoint_delete_handler(JSRuntime* rt, JSInlineCacheRingSlot *ref, uint8_t offset, JSAtom atom, intptr_t target);
static void ic_watchpoint_free_handler(JSRuntime* rt, JSInlineCacheRingSlot *ref, uint8_t offset, JSAtom atom);
```
Since you don't check the return value and I don't see why you'd want to pass `ref` untyped.